esrally/metrics.py

```diff
@@ -1651,6 +1696,112 @@ def index_name(self, race):
         race_timestamp = race.race_timestamp
         return f"{EsRaceStore.INDEX_PREFIX}{race_timestamp:%Y-%m}"
 
+    def add_annotation(self):
+        def _at_midnight(race_timestamp):
+            TIMESTAMP_FMT = "%Y%m%dT%H%S%MZ"
+            date = datetime.datetime.strptime(race_timestamp, TIMESTAMP_FMT)
+            date = date.replace(hour=0, minute=0, second=0, tzinfo=datetime.timezone.utc)
+            return date.strftime(TIMESTAMP_FMT)
+
+        environment = self.environment_name
+        # To line up annotations with chart data points, use midnight of day N as this is
+        # what the chart use too.
+        race_timestamp = _at_midnight(self._race_timestamp())
+        track = self._track()
+        chart_type = self._chart_type()
+        chart_name = self._chart_name()
+        message = self._message()
+        annotation_id = str(uuid.uuid4())
+        dry_run = self._dry_run()
+
+        if dry_run:
+            console.println(
+                f"Would add annotation with message [{message}] for environment=[{environment}], race timestamp=[{race_timestamp}], "
+                f"track=[{track}], chart type=[{chart_type}], chart name=[{chart_name}]"
+            )
+        else:
+            if not self.client.exists(index="rally-annotations"):
+                body = self.index_template_provider.annotations_template()
+                self.client.create_index(index="rally-annotations", body=body)
+            self.client.index(
+                index="rally-annotations",
+                id=annotation_id,
+                item={
+                    "environment": environment,
+                    "race-timestamp": race_timestamp,
+                    "track": track,
+                    "chart": chart_type,
+                    "chart-name": chart_name,
+                    "message": message,
+                },
+            )
+            console.println(f"Successfully added annotation [{annotation_id}].")
+
+    def list_annotations(self):
+        environment = self.environment_name
+        track = self._track()
+        from_date = self._from_date()
+        to_date = self._to_date()
+        query = {
+            "query": {
+                "bool": {
+                    "filter": [
+                        {"term": {"environment": environment}},
+                        {"range": {"race-timestamp": {"gte": from_date, "lte": to_date, "format": "basic_date"}}},
+                    ]
+                }
+            }
+        }
+        if track:
+            query["query"]["bool"]["filter"].append({"term": {"track": track}})
+
+        query["sort"] = [{"race-timestamp": "desc"}, {"track": "asc"}, {"chart": "asc"}]
+        query["size"] = self._max_results()
+
+        result = self.client.search(index="rally-annotations", body=query)
```

Comment:
Not a big deal, but if I run `esrally list annotations` against a metric store that doesn't have an existing index I get a cryptic error on the CLI:

```
$ esrally list annotations

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[ERROR] Cannot list. A transport error occurred while running the operation [search] against your Elasticsearch metrics store on host [localhost] at port [9200].

Getting further help:
*********************
* Check the log files in /Users/bradleydeam/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/bradleydeam/.rally/logs.

-------------------------------
[INFO] FAILURE (took 0 seconds)
-------------------------------
```

I have to dig into the logs to discover why:
```
2022-12-22 06:07:58,777 -not-actor-/PID:37123 elasticsearch WARNING POST https://localhost:9200/rally-annotations/_search [status:404 request:0.002s]
2022-12-22 06:07:58,777 -not-actor-/PID:37123 esrally.metrics ERROR A transport error occurred while running the operation [search] against your Elasticsearch metrics store on host [localhost] at port [9200].
Traceback (most recent call last):
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 118, in guarded
    return target(*args, **kwargs)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/client/utils.py", line 168, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/client/__init__.py", line 1668, in search
    return self.transport.perform_request(
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/transport.py", line 458, in perform_request
    raise e
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/transport.py", line 419, in perform_request
    status, headers_response, data = connection.perform_request(
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 277, in perform_request
    self._raise_error(response.status, raw_data)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/connection/base.py", line 330, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
elasticsearch.exceptions.NotFoundError: NotFoundError(404, 'index_not_found_exception', 'no such index [rally-annotations]', rally-annotations, index_or_alias)
2022-12-22 06:07:58,778 -not-actor-/PID:37123 esrally.rally ERROR Cannot run subcommand [list].
Traceback (most recent call last):
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 118, in guarded
    return target(*args, **kwargs)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/client/utils.py", line 168, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/client/__init__.py", line 1668, in search
    return self.transport.perform_request(
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/transport.py", line 458, in perform_request
    raise e
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/transport.py", line 419, in perform_request
    status, headers_response, data = connection.perform_request(
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 277, in perform_request
    self._raise_error(response.status, raw_data)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/.venv/lib/python3.8/site-packages/elasticsearch/connection/base.py", line 330, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
elasticsearch.exceptions.NotFoundError: NotFoundError(404, 'index_not_found_exception', 'no such index [rally-annotations]', rally-annotations, index_or_alias)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/rally.py", line 1080, in dispatch_sub_command
    dispatch_list(cfg)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/rally.py", line 847, in dispatch_list
    metrics.list_annotations(cfg)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 1249, in list_annotations
    race_store(cfg).list_annotations()
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 1600, in list_annotations
    return self.es_store.list_annotations()
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 1761, in list_annotations
    result = self.client.search(index="rally-annotations", body=query)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 104, in search
    return self.guarded(self._client.search, index=index, body=body)
  File "/Users/bradleydeam/perf/github.com/b-deam/rally/esrally/metrics.py", line 177, in guarded
    raise exceptions.RallyError(msg)
esrally.exceptions.RallyError: A transport error occurred while running the operation [search] against your Elasticsearch metrics store on host [localhost] at port [9200].
``` 

Can we add a catch to print something more helpful, perhaps with a suggestion?

Reply:
good catch, i will update to add an error message

Reply:
It will now print:
```
[ERROR] Cannot list. The operation [search] against your Elasticsearch metrics store on host [98148308b75c45418afe604f7948def1.europe-west1.gcp.cloud.es.io] at port [9243] failed because index [rally-annotations] does not exist.

```